### PR TITLE
feat: Add CommonJS import support via Rollup

### PR DESCRIPTION
**Pull request by Exponential-Workload**
## What is this PR?

To support ESM and CJS, this PR introduces Rollup, a JavaScript Bundler and Preprocessor similar to Webpack or Babel, to both target CJS and ESM, as well as minify the code.

In addition, it introduces a .npmignore file, to avoid shipping tests, bench and src to npm, further reducing the size of the bundle.

## Why?

Some people are stuck in last-generation TS/JS projects that would require too much effort to migrate entirely to ESM - be it due to broken dependencies, or otherwise.

In addition, if a preprocessor is added, it can deliver minified builds to npm, allowing for a lower bundle size

## Dependencies

This introduces 4 new direct development dependencies:
- rollup
  - @rollup/plugin-terser
  - @rollup/plugin-typescript
    - tslib

It does **not** introduce any new production dependencies.

### CODE REVIEW

File: .npmignore
Comment (cyyynthia) [inline .npmignore:1]: This is unnecessary; The `files` property in the package.json explicitly lists the files to push already.

File: rollup.config.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        sourcemap: true,
+      }
+    ],
+    plugins: [
+      typescript({
+        tsconfig: './tsconfig.json',
+        sourceMap: true,
+        declaration: true,
+      }),
+      terser(),
```
Comment (cyyynthia) [inline rollup.config.ts]: I strongly dislike this. I believe packages on NPM should come as free of any pre-processing as possible, besides the strict minimum to get a working package (typescript -> js).

Doing this here makes the dependency completely unhackable for attempts at poking at the code. Many libs are moving to plain JS with JSDoc-based type checking for this very reason (and I might consider moving to that too).

It also makes it extremely difficult to look through code during use and generally (even with source maps unfortunately, speaking from experience) makes errors completely irrelevant due to poor/subpar support.

Comment (Exponential-Workload) [reply rollup.config.ts:25]: > I strongly dislike this. I believe packages on NPM should come as free of any pre-processing as possible, besides the strict minimum to get a working package (typescript -> js).

Personally I find minimal package size is important, however it is your choice at the end of the day



> Doing this here makes the dependency completely unhackable for attempts at poking at the code. Many libs are moving to plain JS with JSDoc-based type checking for this very reason (and I might consider moving to that too).

Manipulating code in node_modules is heavily discouraged by anyone - It's typically encouraged to just fork the source (when available) and publish a new, usually scoped, npm package

> It also makes it extremely difficult to look through code during use and generally (even with source maps unfortunately, speaking from experience) makes errors completely irrelevant due to poor/subpar support.

Fair enough lol

File: .npmignore
Comment (Exponential-Workload) [reply .npmignore:1]: Didn't notice that, my bad

**Comment by Exponential-Workload** [comment]
> An alternate way of doing it can be bundling the lib with tsc (like before) but as CJS by modifying the tsconfig - and make sure the CJS export can be used with ESM imports. This is what [fastify](https://github.com/fastify/fastify/blob/ced4672bd64faeaa84b36c23536bf450c5c919d8/fastify.js#L861-L874) does for example (minus the tsc build since it's written in plain js).

Alright, I'll remove terser from this PR.

If you strongly want to avoid Rollup altogether (despite it having extremely minimal downsides), please close this PR - if I have time I'll open a PR doing it the alternate way you mentioned.

**Comment by Exponential-Workload** [comment]
Also, it may be worth looking into [this StackOverflow post](https://stackoverflow.com/questions/50365177/using-rollup-to-only-convert-to-commonjs-without-bundling#50427393) if you don't mind rollup but simply want to avoid bundling everything into one file.

File: rollup.config.ts (hunk 1 above)
Comment (cyyynthia) [reply rollup.config.ts]: > Manipulating code in node_modules is heavily discouraged by anyone - It's typically encouraged to just fork the source (when available) and publish a new, usually scoped, npm package with said desired changes.

I am well aware of the best practice here, but it remains a handy way to poke at the source code, and for curious souls to lure through it without having to go to a GitHub repository. Minified code can hide all sorts of things, from beautiful code to supply chain issues.

> Personally I find minimal package size is important, however it is your choice at the end of the day.

It is completely irrelevant for a NPM library package, let alone for a package this size.

**Comment by cyyynthia** [comment]
> If you strongly want to avoid Rollup altogether (despite it having extremely minimal downsides)

They're not "extremely minimal". Another hidden cost of bundling a lib like this with more than the strict minimum processing is the including of unwanted polyfills. I think Rollup is relatively tame on this these days, but NPM is full of package pre-bundled with a lot of polyfill code which is absolutely terrible (duplicated polyfill code, or unnecessary polyfill that's slower than the now-supported syntaxes).

It may not seem like it, but some seemingly innocent packaging choices may have negative impact that'll stick for decades.

In addition, this is more of a personal touch, but I generally prefer to keep the amount of tooling I rely on minimal, especially when there are ways to achieve the same thing with much less trouble. This is why I mentioned removing the TS bundle step altogether (like Svelte did a while ago, for similar reasons). I am careful with the tools I bring in my projects (sometimes too much!) and it generally saved me tons of headaches (I still remember all the different bundlers hype before Vite came out on top, with people adopting Snowpack... Anyway, I'm digressing here)

> If I have time I'll open a PR doing it the alternate way you mentioned.

Thank you for your time, and thank you for the PR!! If you don't find the time, I'll take care of bringing CJS support in the future.

I don't want to sound like I'm bashing on your work, haha. It just doesn't fit with the way I'd like the project to be structured. Thanks again for this PR, and let me know how the other way I proposed is going (if you're willing to work on that). 🙌🏻 ❤️

**Pull request closed without merging**